hs_core/serialization.py

```diff
@@ -0,0 +1,842 @@
+
+from hs_core.hydroshare.utils import get_resource_types
+from hs_core.hydroshare.date_util import hs_date_to_datetime, hs_date_to_datetime_iso
+
+from hs_core.hydroshare.utils import resource_pre_create_actions
+from hs_core.hydroshare.utils import ResourceFileSizeException, ResourceFileValidationException
+from hs_core.hydroshare import create_resource
+
+
+class HsSerializationException(Exception):
+    pass
+
+
+class HsDeserializationException(HsSerializationException):
+    pass
+
+
+def _prepare_resource_files_for_creation(file_paths):
+    res_files = []
+
+    for fp in file_paths:
+        fname = os.path.basename(fp)
+        fsize = os.stat(fp).st_size
+        fd = open(fp, 'rb')
+        res_files.append(UploadedFile(file=fd, name=fname, size=fsize))
+
+    return res_files
+
+
+def create_resource_from_bag(bag_content_path, preserve_uuid=True):
+    """
+    Create resource from existing uncompressed BagIt archive.
+
+    Follows roughly the same pattern as hs_core.views.create_resource().
+
+    :param bag_content_path:
+    :return:
+    :raises: HsDeserializationException if an error occurred.
+    """
+    # Get resource metadata
+    resource_files = None
+    try:
+        rm = GenericResourceMeta.read_metadata_from_resource_bag(bag_content_path)
+        res_files = [os.path.join(bag_content_path, f) for f in rm.files]
+        resource_files = _prepare_resource_files_for_creation(res_files)
+
+    except GenericResourceMeta.ResourceMetaException as e:
+        msg = "Error occurred while trying to create resource from bag path {0}. "
+        msg += "Error was: {1}"
+        msg = msg.format(bag_content_path, str(e))
+        raise HsDeserializationException(msg)
+
+    # Send pre-create resource signal
+    try:
+        if resource_files is None:
+            resource_files = []
+        page_url_dict, res_title, metadata = resource_pre_create_actions(resource_type=rm.res_type,
+                                                                         files=resource_files,
+                                                                         resource_title=rm.title,
+                                                                         page_redirect_url_key=None)
+    except ResourceFileSizeException as ex:
+        raise HsDeserializationException(ex.message)
+
+    except ResourceFileValidationException as ex:
+        raise HsDeserializationException(ex.message)
+
+    except Exception as ex:
+        raise HsDeserializationException(ex.message)
+
+    # Create the resource
+    resource = None
+    try:
+        # Get user
+        owner_uri = rm.get_owner_uri().strip()
+        owner_pk = os.path.basename(owner_uri)
+        # Override for testing
+        owner_pk = 1
+
+        resource_id = None
+        if preserve_uuid:
+            resource_id = rm.id
+
+        kwargs = {}
+        if rm.creation_date:
+            # This has no effect
+            kwargs['created'] = rm.creation_date
```

Comment:
@selimnairb I'll defer to @selimnairb for specific suggestions, but as far as I can see, Date metadata element inherited from AbstractmetaDataElement probably is used to record resource creation date, modification date, etc. Perhaps the creation date can be passed in when creating this Date metadata element for the ingested resource.